### ISSUE TITLE
Telegram Desktop for Linux freezes if you load 2 full screen images in a row

### ISSUE THREAD

**Issue by asbestomolesto** [telegramdesktop/tdesktop#17296]
### Steps to reproduce

Click on an image, it shows full screen
exit, you can use the chat etc.
Click on another image - nothing happens, telegram is stuck but it's working in background.

I investigate this and thanks to a friend that helped me, I discovered that the full screen window for the first picture has been mapped once; then unmapped

Then I click the other image and the wm never sees another map request again - telegram does not seem to attempt to map the window a 2nd time, but it behaves as if it was mapped

thus the freeze.

Hope this can help.

### Expected behaviour

It must go back to the chat and continue to be usable :) 

### Actual behaviour

It kinda locks itself and it's not usable anymore. I have to kill and restart it. 

### Operating system

DEVUAN Chimaera (debian based distro) Gnu/Linux, Kernel 5.10.0-9-amd64, Enlightenment 0.24.99.24677 

### Version of Telegram Desktop

3.2.5

### Installation source

Static binary from official website

**Comment by asbestomolesto** [comment]
I found out that 3.1.8 was the last version that worked flawlessy. From version 3.1.10 I can't open a second image after the first one went full screen.

**Comment by ilya-fedin** [comment]
3.1.10 is the first version built with Qt 6. Apparently, a Qt 6 regression? Although, I can't reproduce this, I use NixOS unstable with MATE.

**Comment by asbestomolesto** [comment]
Ok, but what about the wm not having another map request ?  It seem to me that telegram do not map the window a 2nd time, but it behaves as if it was mapped. I think this is not a correct behaviour...

**Comment by ilya-fedin** [comment]
Well, I have no idea, such low-level things like 'mapping the window' are in Qt responsibility, tdeskop uses high-level APIs like QWidget::show and etc.

**Comment by Docbroke** [comment]
I am facing same issue, on my alpine linux edge as well as on archlinux, both x86-64 with cwm window manager.

**Comment by ilya-fedin** [comment]
Oh, well, it's bad. I have no idea how to report this to Qt because:
1. I can't reproduce this. Apparently, it's graphics drivers-dependent.
2. It's not clear what actually happens, media viewer is a pretty coupled widget.
3. Summing previous two points, it's impossible to provide a minimal reproducible example in a Qt bugreport what would lead the report to be closed.

**Comment by asbestomolesto** [comment]
I don't think this is a QT6 issue because, for example, Enlightenment Desktop Manager works well with any QT6 program dealing windows but Telegram :/ 

As I mentioned, problem seem to be Telegram not having another map request - it doesn't map the window a 2nd time, but it behaves as if it was mapped. Hence the problem. 

It just need to map the window all the times.

**Comment by ilya-fedin** [comment]
What do you mean by 'map request'?

**Comment by ilya-fedin** [comment]
Media viewer doesn't show maps AFAIK

**Comment by ilya-fedin** [comment]
> Enlightenment Desktop Manager works well with any QT6 program dealing with windows opening and closing, but Telegram :/

I believe tdesktop due to its architecture can trigger Qt bugs in such operations even where other apps seem to work. Media viewer is a very advanced widget that exists all the runtime of tdesktop and uses OpenGL. I believe your other Qt 6 applications don't use APIs tdesktop uses.

**Comment by rastermann** [comment]
its a client-side bug. qt probably but from x11/the window manager's point of view it doesn't matter - it's an x11 client that talks to the xserver (and via the xserver indirectly to the window manager). telegram withdraws the window and tries to re-use it (as opposed to destroys the window and creates a new one). please see the icccm manual section 4: https://tronche.com/gui/x/icccm/sec-4.html.

4.1.4. Changing Window State
Normal -> Withdrawn - The client should unmap the window and follow it with a synthetic UnmapNotify event as described later in this section.

the client is responsible for removing WM_STATE. the client does NOT remove this when the window is withdrawn - it is still there. the client (telegram) withdraws the window with an unmapnotify event. that is what E (the wm) sees. E destroys all of its associated objects surrounding that window at this point as the client withdrew. e does NOT remove WM_STATE because the client did the withdraw. the client never tries to map the window again. read the icccm to know what map/unmap are. (they are show/hide - x11 terminology that has been around for 30-40 years). new windows when shown are mapped by a client. wm's set up a redirect on the root window that turns a map into a maprequest event inside the xserver and this is sent to the wm. wm's then know some window wants to be mapped and they go handle this. this is one of the core things a wm does and how a lot of this even works to begin with. e never sees another maprequest from telegram after you close this media window. that means the client (telegram) never maps the window again.

i poked around and qt/telegram (the client) refuse to map the window as long as the WM_STATE property is there on this media window. if i manually set this WM_STATE property to withdrawn, it still wont map the window. i have to actively remove the property (i can do this from outside the app - any x client can mess with any other x client's properties on their window). this screams to me of qt not looking at the mapped/unmapped state of the window but instead only looking at the WM_STATE property only. the mapped/unmapped state is the primary indicator of window "visibility" not the WM_STATE property. WM_STATE exists to add fine-grained info if the window is mapped normally, like is it fully visible or is it in an iconified state (so eg maybe minimized to taskbar and maybe just shown as a thumbnail etc.).

yes - further in the icccm:

4.1.3.1. WM_STATE Property
When the window is withdrawn, the window manager will either change the state field's value to WithdrawnState or it will remove the WM_STATE property entirely.

the client should already have done this as per icccm, but as above - if i manually change the WM_STATE to withdrawn the window still will never show again - the property must be completely removed. if e were to do the above and set it to withdrawn qt/telegram (the client) would still be at fault as it would still not work.

so congratulations... you have here a low level telegram bug or a qt bug. :) i've already debugged as far as the above... solutions could be to ... find the bug in qt and fix qt, don't unmap the media window but destroy it completely and re-create it each time (this works around the qt bug). i can work around it in e and forcibly remove WM_STATE ... but workarounds don't fix the core bug. :)

**Comment by asbestomolesto** [comment]
I had a chat with "raster", the main developer of Enlightenment. Here's what he says about this bug: 


<asbesto> Just tried to update telegram for linux again, and that bug is still there :( This bug: https://github.com/telegramdesktop/tdesktop/issues/17296
<asbesto> raster: can you please check that github thread so maybe you can figure out something? :( 
<asbesto> I have this bug only with Enlightenment :( 
<raster> maprequest is an x11 thing
<raster> mapping a window == XMapWindow - show a window
<raster> wm;'s set up an indirtection on root to intercept maps
<raster> and wm's get a maprequest event if an app tries to map a window
<raster> the window does not get mapped
<raster> the wm gets that event
<raster> the wm then can take the window and go map it...
<raster> thats how wm's work
<raster> its a built in mechanism in x to become a wm
<raster> (among other things)
<raster> qt seems to not care the window is not mapped
<raster> the window shows
<raster> xwininfo says
<raster>   Map State: IsViewable
<raster> just run xwininfo on it
<raster> well run xwininfo and clikc on it
<raster> when the window hides it is
<raster>   Map State: IsUnMapped
<raster> eg i do xwininfo -id 0x200001c
<raster> (0x2... is the id of the window as u managed ot see the firts time u ran xwininfo and clikce don it)
<raster> it's unmapped
<raster> telegram refuses to try map the window ever agin
<raster> it doesnt try show it
<raster> e sees no  maprequest
<raster> asbesto: and i confirmed. its unmaped - but as long as it has a WM_STATE property on wit ... which is nbot the primary source of visibility - the window refuses to re-map
<raster> clients should rely on the window map state not the property
<raster> ththe property is only useful if the window is still mapped
<raster> to know if its iconified or fully visible
<raster> it is not a primary property of visibility - the map state is
<raster> a wm could crash or be replaced - something else unmap the window
<raster> and the property may not have any state that is useful because the mapped state of the window is the primary thing to worry about

<raster> i can work around this in e
<raster> but relying on WM_STATE and not using the mapped state of a window to know if its visible or not is just asking for trouble
<raster> as above - lots of situations where they may not be in sync

<raster> in fact the icccm is super clear
<raster> 4.1.4. Changing Window State
<raster> Normal -> Withdrawn - The client should unmap the window and follow it with a synthetic UnmapNotify event as described later in this section.
<raster> that is what actually happens
<raster> if i add debug to e
<raster> X HIDE: 0x320001c
<raster> thats in
<raster> _e_comp_x_hide() which handles window hides (unmapnotify events)
<raster> and under
<raster> 4.1.3.1. WM_STATE Property
<raster> he window manager will place a WM_STATE property (of type WM_STATE) on each top-level client window that is not in the Withdrawn state. Top-level windows in the Withdrawn state may or may not have the WM_STATE property. Once the top-level window has been withdrawn, the client may re-use it for another purpose. Clients that do so should remove the WM_STATE property if it is still present.
<raster> qt or telegram do NOT remove WM_STATE
<raster> e doesnt remove it either
<raster> its telegram that withdrew the window
<raster> (or qt - -the client - not the wm)
<raster> the client should remove WM_STATe
<raster> qt REFUSES to map the window as long as WM_STARE is there
<raster> the WM_STTE ie REFUSED to remove to begin with
<raster> icccm says the client should od this if it wants to re-use the window
<raster> its trying to re-use it
<raster> so sure
<raster> i can work around this and remove the wm state for the app
<raster> but icccm says it's the client's job...
<raster> fyi if i change the wm_state to withdrawn
<raster> and keep the prop there
<raster> the window still doesnt show
<raster> so later on on that age
<raster> When the window is withdrawn, the window manager will either change the state field's value to WithdrawnState or it will remove the WM_STATE property entirely.
<raster> won't help
<raster> as qt insists ythe property must be totally removed
<raster> but it doesnt remove it...
<raster> i can work around it... but its some qt/telegram bug 
<raster> as per icccm standards :)

**Comment by ilya-fedin** [comment]
Guys, you seem to know better what is needed here, so it would a way faster if you would report this to Qt directly

**Comment by rastermann** [comment]
I'll leave that to @asbestomolesto :) but as above - you may want to work around qt and destroy the window not just hide it...

**Comment by ilya-fedin** [comment]
It's hidden on a reason: it's very heavy and its construction takes a lot of time, so it's created before main window is shown and exists all the tdesktop run time.

**Comment by rastermann** [comment]
it can't be THAT expensive to create a window... certainly with content that simple (just show an image/video/whatever and some close buttons and other controls)... like a few dozen ms on a bad day...

at least with elementary i measure about 4ms to create a new video player window with a video object content that plays streams and files on my desktop for an additional window (the first window takes about 50ms as it pays all the initial costs of loading theme information from files that are then re-used later).

**Comment by ilya-fedin** [comment]
> it can't be THAT expensive to create a window...

window is created indirectly by Qt, what tdesktop creates is a QWidget that is indeed heavy to create in the media viewer case.

**Comment by ilya-fedin** [comment]
btw, on my machine xprop says that WM_STATE is Withdrawn and xwininfo says that the window InUnMapped, but as soon as I click at any media, media viewer opens, it's map state changes to IsViewable and WM_STATE changes to Normal. Am I missing something?

**Comment by rastermann** [comment]
i actually forcibly set WM_STATE to Withdrawn and qt refused to map the window again. it stayed unmapped. (i wrote a small few lines of code that went to the window id of the media window of telegram and forced it to withdrawn or could also remove the WM_STATE property). qt would ONLY map the window if WM_STATE was removed from the window. it would otherwise not show the window. Qt was relying on the presence of this property to decide if the window should be mapped or not (seemingly). As above - it's the client's job to remove WM_STATE if the client withdraws (unmaps) the window. if Qt didn't rely on WM_STATE it wouldn't matter as Qt would map the window since it's unmapped ... but as Qt refuses to map it if WM_STATE exists ... then ... in circles. :)

**Comment by ilya-fedin** [comment]
I don't see any check for WM_STATE in Qt code. It just calls xcb_map_window, there's not early return anywhere. So maybe xcb_map_window doesn't work on your machine for some reason.
https://github.com/qt/qtbase/blob/v6.2.3/src/plugins/platforms/xcb/qxcbwindow.cpp#L678-L729

**Comment by ilya-fedin** [comment]
Or maybe it's an Enlightement bug
https://github.com/qt/qtbase/blob/v6.2.3/src/plugins/platforms/xcb/qxcbwindow.cpp#L2232-L2236

**Comment by rastermann** [comment]
as above - i wrote a tool to literally remove or modify WM_STATE and the window would not show again until i remove WM_STATe from the window manually with that tool whilst the window is unmapped ... then it shows. IF Qt is waiting for the WM to remove WM_STATE... then it's wrong. I pointed to the ICCCM manual above and quoted it.

4.1.3.1. WM_STATE Property ... The window manager will place a WM_STATE property (of type WM_STATE) on each top-level client window that is not in the Withdrawn state. Top-level windows in the Withdrawn state may or may not have the WM_STATE property. Once the top-level window has been withdrawn, the client may re-use it for another purpose. Clients that do so should remove the WM_STATE property if it is still present.

This is what Qt does - re-use the window. Qt withdraws the window. it's probably waiting for WM_STATE to change but the client should remove WM_STATE if still there if it wishes to re-use it. Qt waiting on this from the WM when they want to re-use the window is specifically violating the above section of the ICCCM. So literally the code you point to in Qt is actually wrong. if that deferred logic will not re-map the window until WM_STATE has been removed... which is exactly what i see here as above (even if WM_STATE is set to withdrawn then it still wont get mapped - that is what i saw when modifying or removing WM_STATE on this window).

**Comment by ilya-fedin** [comment]
It doesn't wait for WM_STATE to be removed, I don't see such code and I can confirm that on my machine media viewer opens just fine even with WM_STATE=Withdrawn. All it does is waits for a property notify event as ICCCM says and it seems Enlightenment doesn't send such event violating ICCCM as it works on MATE and it doesn't remove WM_STATE.

**Comment by rastermann** [comment]
4.1.3.1 says the client should remove WM_STATE if it is to be re-used. i quoted it above right there. that was copy & paste from ICCCM. and as i mentioned i set wm state to withdrawn and the window still didn't show - only when entirely removed did it work. e sees no maprequests until then. maprequests are generated by the xserver when a client tries to map a window that is an immediate child of the root window.

**Comment by ilya-fedin** [comment]
Ok, feel free to report to Qt if you think Qt is at fault

**Comment by smabie** [comment]
Any update on this? Running telegram 3.6.1 on cwm as well and this bug makes telegram virtually unusable.

**Comment by ilya-fedin** [comment]
No, I still can't reproduce this. Runs just fine on my machine.

**Comment by smabie** [comment]
> No, I still can't reproduce this. Runs just fine on my machine.

Maybe you could try running telegram-desktop under cwm to debug?

**Comment by ilya-fedin** [comment]
what's cwm?

**Comment by smabie** [comment]
https://github.com/leahneukirchen/cwm

https://en.wikipedia.org/wiki/Cwm_(window_manager)

**Comment by ilya-fedin** [comment]
Oh, I don't see titlebars in Wikipedia screenshot and I don't know how to work with such kind of window managers, so I would be unable to do this.

**Comment by smabie** [comment]
> Oh, I don't see titlebars in Wikipedia screenshot and I don't know how to work with such kind of window managers, so I would be unable to do this.

You could spend 5 minutes reading the documentation?

**Comment by ilya-fedin** [comment]
> You could spend 5 minutes reading the documentation?

Well, I'm not really interested in doing that 😅

**Comment by foodornt** [comment]
This has been out there for almost a year, I could've learned CPP and QT and solved this thing with a pull request

**Comment by ilya-fedin** [comment]
Please do. A fix from someone who can reproduce this & interested in fixing is something this issue really needs.

**Comment by ilya-fedin** [comment]
You would also need to learn libxcb & X11 protocol though as this seem to be a bug in how Qt works with X11.

**Comment by foodornt** [comment]
Will do, hope my 16gb of ram can survive building tdesktop

**Comment by Docbroke** [comment]
The issue is no longer there in telegram-desktop 3.7.3. It appears to be solved.
